README.md
# PHP Module Player

The PHP Module Player is an extension for the PHP language written in pure C code, built on top of the libmikmod that is linked as a submodule in this repository.

The intent of this extension is to enabled the support to play [tracker music](https://en.wikipedia.org/wiki/Module_file), also know as chiptune songs, from the PHP CLI SAPI.

The extension supports all tracker related extensios, like .MOD, .XM, .IT, .S3M and etc. For more format support information, check the [libmikmod](http://mikmod.sourceforge.net/) documentation.

Here is a [video](https://www.youtube.com/watch?v=rSjuUIpTbYk) demonstrating the usage of the extension.

> I made this extension just for fun and research and I couldn't support it in full-time, so, if you find any bug or need to make a contribution, feel free to send me a PR.

# Requirements

- PHP 7 (for older versions, checkout to the git tag 1.1.0)
- libmikmod (included in the repository)

# Installation

As this extension is not really useful for everyone, it will not be available in the PHP repository for easy installation, but the process is quite simple.

### 1. Clone the repository

The first thing you need to do, is to clone this repository and initialize it submodule:

```
$ git clone git@github.com:devsdmf/php-mod-player.git /tmp/path/to/extension
$ cd /tmp/path/to/extension
$ git submodule update --init
```

### 2. Installing the libmikmod

This extension requires the libmikmod to be installed and located in the system path, se we need to compile and install it:

> To compile the library, there is a lot of ways, here I'll demonstrate the simplest way, but if you have any problem, you could check the library documentation or ask me.

```
# assuming that you're in the repository root
$ cd mikmod/libmikmod
$ ./configure
$ make
$ sudo make install
```

> If the configure command fails, you could recompile the autoconf script using the following commands in the library root.
```
$ autoreconf --install && autoconf configure.ac > configure
$ chmod +x configure
```
Now, try to compile the library again as describe above.

### 3. Installing the extension

Now, we just need to compile the extension as any other PHP extension.

```
# assuming that you're in the repository root
$ phpize
$ ./configure
$ make
$ sudo make install
```

> You may need to add the path to the modplayer.so file in your php.ini file.

And it's done!

# Usage

The extension itself, has only three functions, all of them is described below:

### modplayer_play

(PHP 5 >= 5.5, PHP 7)
modplayer_play - Play a module file from the disk

#### Description
`int modplayer_play( string $filename [, int max_channels [, bool curious [, int reverb ]]] )`

Start the module player and play the $filename file in background.

#### Parameters

- *$filename* - The module audio file in the folder, or with the absolute path.
- *$max_channels* - The tracker channels to be used (default: 64).
- *$curious* - Sets the mixer to search and mix the audio on the fly (default: false).
- *$reverb* - Sets the reverb level from 0 to 15 (default: 1).

#### Return Values

The function returns the PID of the player sub-process or FALSE if an error occurs.

#### Examples

Playing a sample module file:

```php
if (!modplayer_getpid()) {
    $pid = modplayer_play('sample.mod');
} else {
    echo 'Module already playing.';
}
```

### modplayer_getpid

(PHP 5>= 5.5, PHP 7)
modplayer_getpid - Get the PID of the player subprocess

#### Description
`int modplayer_getpid( )`

Get the PID of the player subprocess.

#### Parameters

N/A

#### Return Values

The function returns the PID of the player sub-process or FALSE if none is playing.

#### Examples

Getting the PID of the player sub-process:

```php
$pid = modplayer_getpid();
```

### modplayer_stop

(PHP 5>= 5.5, PHP 7)
modplayer_stop - Stop a currently playing song.

#### Description
`bool modplayer_stop( )`

Stop a currently playing song.

#### Parameters

N/A

#### Return Values

Returns **TRUE** if the process was stopped otherwise **FALSE** if an error occurs.

#### Examples

Stopping the player

```php
if (modplayer_getpid()) {
    if (modplayer_stop()) {
        echo "Success!";
    } else {
        echo "An error occured!";
    }
}
```

# License

This library is licensed under the [MIT license](LICENSE).

CHANGELOG.md
# CHANGELOG

2.0.0
-----

- Updated the libmikmod to its latest version (3.3.12)
- PHP 7.0 and 7.1 support
- Changed INI configurations to function signature parameters (max channels and curious)
- Added reverb effect
- Improved stability and sound quality

1.1.0
-----

- Changed the extension name from *cli_audio_stream* to *modplayer*
- Bug fixes
- Standard fixes

1.0.0
-----

- First public release for PHP 5.5 and 5.6


stream.c
#include <sys/types.h>
#include <unistd.h>
#include <stdlib.h>
#include "php.h"
#include "stream.h"
#include "mikmod.h"

int stream_audio(FILE *file, int maxchan, int curious)
{
	signed int s_pid = 0;
	pid_t pid = fork();

	s_pid = getpid();

	if (pid == 0) {
		MODULE *module;

		MikMod_InitThreads();
        MikMod_RegisterAllDrivers();
        MikMod_RegisterAllLoaders();

        md_mode |= DMODE_SOFT_MUSIC | DMODE_NOISEREDUCTION | DMODE_INTERP;
        if (MikMod_Init("")) {
            // fprintf(stderr, "Could not initialize sound, reason: %s\n",MikMod_strerror(MikMod_errno));
            zend_error(E_ERROR, "Could not initialize the MikMod library");
            return;
        }

        module = Player_LoadFP(file, maxchan, curious);
        if (module) {
        	module->wrap = 1;
        	module->loop = 0;

        	Player_Start(module);

        	while (Player_Active()) {
        		usleep(10000);
        		MikMod_Update();
        	}
        } else {
        	zend_error(E_ERROR, "Could not load module");
        }

        MikMod_Exit();
	} else if (pid < 0) {
		zend_error(E_ERROR, "Failed to fork CLI Audio Stream process");
		exit(1);
	}

	return s_pid;
}

cli_audio.c
/**
 * PHP CLI AUDIO STREAMMER
 *
 * This is a extension to provide a stream of module audio files under
 * the CLI SAPI. This library is licensed under the MIT license.
 * 
 * Developed by: devSDMF <devsdmf@gmail.com>
 *
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <unistd.h>
#include <signal.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/wait.h>

#include "php.h"
#include "php_ini.h"
#include "php_cli_audio.h"
#include "mikmod.h"

ZEND_DECLARE_MODULE_GLOBALS(cli_audio)

static zend_function_entry cli_audio_functions[] = {
    PHP_FE(cli_audio_stream, NULL)
    PHP_FE(cli_audio_getpid, NULL)
    PHP_FE(cli_audio_stop, NULL)
    {NULL, NULL, NULL}
};

zend_module_entry cli_audio_module_entry = {
    #if ZEND_MODULE_API_NO >= 20010901
    STANDARD_MODULE_HEADER,
    #endif
    PHP_CLI_AUDIO_EXTNAME,
    cli_audio_functions,
    PHP_MINIT(cli_audio),
    PHP_MSHUTDOWN(cli_audio),
    NULL,
    NULL,
    NULL,
    #if ZEND_MODULE_API_NO >= 20010901
    PHP_CLI_AUDIO_VERSION,
    #endif
    STANDARD_MODULE_PROPERTIES
};

#ifdef COMPILE_DL_CLI_AUDIO
ZEND_GET_MODULE(cli_audio)
#endif

PHP_INI_BEGIN()
PHP_INI_ENTRY("cli_audio.maxchan","64",PHP_INI_ALL,NULL)
PHP_INI_ENTRY("cli_audio.curious","0",PHP_INI_ALL,NULL)
PHP_INI_END()

static void php_cli_audio_init_globals(zend_cli_audio_globals *cli_audio_globals)
{
    cli_audio_globals->pid = 0;
}

PHP_MINIT_FUNCTION(cli_audio)
{
    ZEND_INIT_MODULE_GLOBALS(cli_audio, php_cli_audio_init_globals, NULL);
    REGISTER_INI_ENTRIES();
    return SUCCESS;
}

PHP_MSHUTDOWN_FUNCTION(cli_audio)
{
    if (CLI_AUDIO_G(pid) > 0) {
        kill(CLI_AUDIO_G(pid), SIGKILL);
    }
    
    UNREGISTER_INI_ENTRIES();
    return SUCCESS;
}

PHP_FUNCTION(cli_audio_stream)
{
    FILE *fptr;
    char *filename;
    int filename_length, s_pid;

    if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &filename, &filename_length) == FAILURE) {
        RETURN_NULL();
    }

    if (CLI_AUDIO_G(pid) > 0) {
        zend_error(E_ERROR, "You've already started the CLI audio stream");
        RETURN_NULL();
    }

    fptr = fopen(filename, "rb");
    if (fptr == NULL) {
        perror("Error");
        zend_error(E_ERROR, "An error occurred at try to open the audio file specified");
        RETURN_NULL();
    }

    s_pid = stream_audio(fptr, INI_INT("cli_audio.maxchan"), INI_INT("cli_audio.curious"));

    if (s_pid > 0) {
        CLI_AUDIO_G(pid) = s_pid;
    }

    RETURN_LONG(s_pid);
}

PHP_FUNCTION(cli_audio_getpid)
{
    RETURN_LONG(CLI_AUDIO_G(pid));
}

PHP_FUNCTION(cli_audio_stop)
{
    if (CLI_AUDIO_G(pid) > 0) {
        kill(CLI_AUDIO_G(pid), SIGKILL);
        CLI_AUDIO_G(pid) = 0;
    }

    RETURN_NULL();
}

int stream_audio(FILE *fptr, int maxchan, int curious)
{
    pid_t m_pid;
    signed int s_pid;

    m_pid = fork();
    if (m_pid == 0) {
        MODULE *module;

        MikMod_InitThreads();
        MikMod_RegisterAllDrivers();
        MikMod_RegisterAllLoaders();

        md_mode |= DMODE_SOFT_MUSIC | DMODE_NOISEREDUCTION | DMODE_INTERP;
        if (MikMod_Init("")) {
            zend_error(E_ERROR, "Could not initialize the MikMod library");
            return -1;
        }

        module = Player_LoadFP(fptr, maxchan, curious);
        if (module) {
            module->wrap = 1;
            module->loop = 0;

            Player_Start(module);

            s_pid = getpid();
            CLI_AUDIO_G(pid) = s_pid;

            while (Player_Active()) {
                usleep(10000);
                MikMod_Update();
            }
        } else {
            zend_error(E_ERROR, "Could not load module");
            return -1;
        }

        fclose(fptr);
        MikMod_Exit();
    } else if (m_pid < 0) {
        zend_error(E_ERROR, "Failed to fork CLI Audio Stream process");
        return -1;
    }

    return m_pid;
}

mod_player.c
/**
 * PHP CLI AUDIO STREAMMER
 *
 * This is a extension to provide a stream of module audio files under
 * the CLI SAPI. This library is licensed under the MIT license.
 * 
 * Developed by: devSDMF <devsdmf@gmail.com>
 *
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <unistd.h>
#include <signal.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/wait.h>

#include "php.h"
#include "php_ini.h"
#include "php_mod_player.h"
#include "mikmod.h"

ZEND_DECLARE_MODULE_GLOBALS(mod_player)

static zend_function_entry mod_player_functions[] = {
    PHP_FE(play_module_file, NULL)
    PHP_FE(mod_player_getpid, NULL)
    PHP_FE(stop_module_file, NULL)
    {NULL, NULL, NULL}
};

zend_module_entry mod_player_module_entry = {
    #if ZEND_MODULE_API_NO >= 20010901
    STANDARD_MODULE_HEADER,
    #endif
    PHP_MOD_PLAYER_EXTNAME,
    mod_player_functions,
    PHP_MINIT(mod_player),
    PHP_MSHUTDOWN(mod_player),
    NULL,
    NULL,
    NULL,
    #if ZEND_MODULE_API_NO >= 20010901
    PHP_MOD_PLAYER_VERSION,
    #endif
    STANDARD_MODULE_PROPERTIES
};

#ifdef COMPILE_DL_MOD_PLAYER
ZEND_GET_MODULE(mod_player)
#endif

PHP_INI_BEGIN()
PHP_INI_ENTRY("mod_player.maxchan","64",PHP_INI_ALL,NULL)
PHP_INI_ENTRY("mod_player.curious","0",PHP_INI_ALL,NULL)
PHP_INI_END()

static void php_mod_player_init_globals(zend_mod_player_globals *mod_player_globals)
{
    mod_player_globals->pid = 0;
}

PHP_MINIT_FUNCTION(mod_player)
{
    ZEND_INIT_MODULE_GLOBALS(mod_player, php_mod_player_init_globals, NULL);
    REGISTER_INI_ENTRIES();
    return SUCCESS;
}

PHP_MSHUTDOWN_FUNCTION(mod_player)
{
    if (MOD_PLAYER_G(pid) > 0) {
        kill(MOD_PLAYER_G(pid), SIGKILL);
    }
    
    UNREGISTER_INI_ENTRIES();
    return SUCCESS;
}

PHP_FUNCTION(play_module_file)
{
    FILE *fptr;
    char *filename;
    int filename_length, s_pid;

    if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &filename, &filename_length) == FAILURE) {
        RETURN_NULL();
    }

    if (MOD_PLAYER_G(pid) > 0) {
        zend_error(E_ERROR, "You've already started the module player");
        RETURN_NULL();
    }

    fptr = fopen(filename, "rb");
    if (fptr == NULL) {
        perror("Error");
        zend_error(E_ERROR, "An error occurred at try to open the module audio file specified");
        RETURN_NULL();
    }

    s_pid = stream_audio(fptr, INI_INT("mod_player.maxchan"), INI_INT("mod_player.curious"));

    if (s_pid > 0) {
        MOD_PLAYER_G(pid) = s_pid;
    }

    RETURN_LONG(s_pid);
}

PHP_FUNCTION(mod_player_getpid)
{
    RETURN_LONG(MOD_PLAYER_G(pid));
}

PHP_FUNCTION(stop_module_file)
{
    if (MOD_PLAYER_G(pid) > 0) {
        kill(MOD_PLAYER_G(pid), SIGKILL);
        MOD_PLAYER_G(pid) = 0;
    }

    RETURN_NULL();
}

int stream_audio(FILE *fptr, int maxchan, int curious)
{
    pid_t m_pid;
    signed int s_pid;

    m_pid = fork();
    if (m_pid == 0) {
        MODULE *module;

        MikMod_InitThreads();
        MikMod_RegisterAllDrivers();
        MikMod_RegisterAllLoaders();

        md_mode |= DMODE_SOFT_MUSIC | DMODE_NOISEREDUCTION | DMODE_INTERP;
        if (MikMod_Init("")) {
            zend_error(E_ERROR, "Could not initialize the MikMod library");
            return -1;
        }

        module = Player_LoadFP(fptr, maxchan, curious);
        if (module) {
            module->wrap = 1;
            module->loop = 0;

            Player_Start(module);

            s_pid = getpid();
            MOD_PLAYER_G(pid) = s_pid;

            while (Player_Active()) {
                usleep(10000);
                MikMod_Update();
            }
        } else {
            zend_error(E_ERROR, "Could not load module");
            return -1;
        }

        fclose(fptr);
        MikMod_Exit();
    } else if (m_pid < 0) {
        zend_error(E_ERROR, "Failed to fork CLI Audio Stream process");
        return -1;
    }

    return m_pid;
}

php_cli_audio.h
/**
 * PHP CLI AUDIO STREAMMER
 *
 * This is a extension to provide a stream of module audio files under
 * the CLI SAPI. This library is licensed under the MIT license.
 * 
 * Developed by: devSDMF <devsdmf@gmail.com>
 *
 */
#ifndef PHP_CLI_AUDIO_H
#define PHP_CLI_AUDIO_H 1

#ifdef ZTS
#include "TSRM.h"
#endif

ZEND_BEGIN_MODULE_GLOBALS(cli_audio)
long pid;
ZEND_END_MODULE_GLOBALS(cli_audio)

#ifdef ZTS
#define CLI_AUDIO_G(v) TSRM(cli_audio_globals_id, zend_cli_audio_globals *, v)
#else
#define CLI_AUDIO_G(v) (cli_audio_globals.v)
#endif

#define PHP_CLI_AUDIO_EXTNAME "cli_audio"
#define PHP_CLI_AUDIO_VERSION "1.0.0"

PHP_MINIT_FUNCTION(cli_audio);
PHP_MSHUTDOWN_FUNCTION(cli_audio);

PHP_FUNCTION(cli_audio_stream);
PHP_FUNCTION(cli_audio_getpid);
PHP_FUNCTION(cli_audio_stop);

int stream_audio(FILE *fptr, int maxchan, int curious);

extern zend_module_entry cli_audio_module_entry;
#define phpext_cli_audio_ptr &cli_audio_module_entry;

#endif

php_modplayer.h
/**
 * PHP CLI AUDIO STREAMMER
 *
 * This is a extension to provide a stream of module audio files under
 * the CLI SAPI. This library is licensed under the MIT license.
 * 
 * Developed by: devSDMF <devsdmf@gmail.com>
 *
 */
#ifndef PHP_MODPLAYER_H
#define PHP_MODPLAYER_H

extern zend_module_entry modplayer_module_entry;

#define phpext_mod_player_ptr &modplayer_module_entry;

#ifdef ZTS
#include "TSRM.h"
#endif

#define PHP_MODPLAYER_EXTNAME "modplayer"
#define PHP_MODPLAYER_VERSION "1.0.0"

ZEND_BEGIN_MODULE_GLOBALS(modplayer)
long pid;
ZEND_END_MODULE_GLOBALS(modplayer)

#ifdef ZTS
#define MODPLAYER_G(v) TSRM(modplayer_globals_id, zend_modplayer_globals *, v)
#else
#define MODPLAYER_G(v) (modplayer_globals.v)
#endif

PHP_MINIT_FUNCTION(modplayer);
PHP_MSHUTDOWN_FUNCTION(modplayer);

PHP_FUNCTION(modplayer_play);
PHP_FUNCTION(modplayer_getpid);
PHP_FUNCTION(modplayer_stop);

#endif

php_mod_player.h
/**
 * PHP CLI AUDIO STREAMMER
 *
 * This is a extension to provide a stream of module audio files under
 * the CLI SAPI. This library is licensed under the MIT license.
 * 
 * Developed by: devSDMF <devsdmf@gmail.com>
 *
 */
#ifndef PHP_MOD_PLAYER_H
#define PHP_MOD_PLAYER_H 1

#ifdef ZTS
#include "TSRM.h"
#endif

ZEND_BEGIN_MODULE_GLOBALS(mod_player)
long pid;
ZEND_END_MODULE_GLOBALS(mod_player)

#ifdef ZTS
#define MOD_PLAYER_G(v) TSRM(mod_player_globals_id, zend_mod_player_globals *, v)
#else
#define MOD_PLAYER_G(v) (mod_player_globals.v)
#endif

#define PHP_MOD_PLAYER_EXTNAME "mod_player"
#define PHP_MOD_PLAYER_VERSION "1.1.0"

PHP_MINIT_FUNCTION(mod_player);
PHP_MSHUTDOWN_FUNCTION(mod_player);

PHP_FUNCTION(play_module_file);
PHP_FUNCTION(mod_player_getpid);
PHP_FUNCTION(stop_module_file);

int stream_audio(FILE *fptr, int maxchan, int curious);

extern zend_module_entry mod_player_module_entry;
#define phpext_mod_player_ptr &mod_player_module_entry;

#endif

modplayer.c
/**
 * PHP CLI AUDIO STREAMMER
 *
 * This is a extension to provide a stream of module audio files under
 * the CLI SAPI. This library is licensed under the MIT license.
 * 
 * Developed by: devSDMF <devsdmf@gmail.com>
 *
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>

#include "php.h"
#include "php_modplayer.h"
#include "mikmod.h"

ZEND_DECLARE_MODULE_GLOBALS(modplayer)

static zend_function_entry modplayer_functions[] = {
    PHP_FE(modplayer_play, NULL)
    PHP_FE(modplayer_getpid, NULL)
    PHP_FE(modplayer_stop, NULL)
    {NULL, NULL, NULL}
};

zend_module_entry modplayer_module_entry = {
    STANDARD_MODULE_HEADER,
    PHP_MODPLAYER_EXTNAME,
    modplayer_functions,
    PHP_MINIT(modplayer),
    PHP_MSHUTDOWN(modplayer),
    NULL,
    NULL,
    NULL,
    PHP_MODPLAYER_VERSION,
    STANDARD_MODULE_PROPERTIES
};

#ifdef COMPILE_DL_MODPLAYER
ZEND_GET_MODULE(modplayer)
#endif

static void php_modplayer_init_globals(zend_modplayer_globals *modplayer_globals)
{
    modplayer_globals->pid = 0;
}

PHP_MINIT_FUNCTION(modplayer)
{
    ZEND_INIT_MODULE_GLOBALS(modplayer, php_modplayer_init_globals, NULL);
    return SUCCESS;
}

PHP_MSHUTDOWN_FUNCTION(modplayer)
{
    if (MODPLAYER_G(pid) > 0) {
        kill(MODPLAYER_G(pid), SIGKILL);
    }

    return SUCCESS;
}

static int play_audio(FILE *fptr, int maxchan, int curious, int reverb)
{
    pid_t m_pid;
    signed int s_pid;

    m_pid = fork();
    if (m_pid == 0) {
        MODULE *module;

        MikMod_InitThreads();
        MikMod_RegisterAllDrivers();
        MikMod_RegisterAllLoaders();

        md_mode |= DMODE_SOFT_MUSIC | DMODE_NOISEREDUCTION | DMODE_INTERP;
        
        if (reverb > 0) {
            md_reverb = reverb;
        }
        if (MikMod_Init("")) {
            php_error_docref(NULL, E_ERROR, "Could not initialize the MikMod library");
            return -1;
        }

        module = Player_LoadFP(fptr, maxchan, curious);
        if (module) {
            module->wrap = 1;
            module->loop = 0;

            Player_Start(module);

            s_pid = getpid();
            MODPLAYER_G(pid) = s_pid;

            while (Player_Active()) {
                usleep(10000);
                MikMod_Update();
            }
        } else {
            fclose(fptr);
            php_error_docref(NULL, E_ERROR, "Could not load module");
            return -1;
        }

        MikMod_Exit();
    } else if (m_pid < 0) {
        php_error_docref(NULL, E_ERROR, "Failed to fork CLI audio stream process");
        return -1;
    }

    return m_pid;
}

/* {{{ proto int modplayer_play(string file [, int reverb [, int max_channels [, bool curious]]])
   Play a module file from the disk and returns the PID of the player sub-process */
PHP_FUNCTION(modplayer_play)
{
    FILE *fptr;
    char *filename, resolved_path[MAXPATHLEN + 1];
    size_t filename_length;
    zend_long max_channels = 64;
    zend_bool curious = 0;
    zend_long reverb = 0;
    int s_pid;

    if (MODPLAYER_G(pid) > 0) {
        php_error_docref(NULL, E_ERROR, "You've already started the module player");
        RETURN_FALSE;
    }

    ZEND_PARSE_PARAMETERS_START(1, 4)
        Z_PARAM_PATH(filename, filename_length)
        Z_PARAM_OPTIONAL
        Z_PARAM_LONG(reverb)
        Z_PARAM_LONG(max_channels)
        Z_PARAM_BOOL(curious)
    ZEND_PARSE_PARAMETERS_END();

    if (strlen(filename) == 0) {
        php_error_docref(NULL, E_ERROR, "You must specify a valid file name");
        RETURN_FALSE;
    }

    if (!expand_filepath(filename, resolved_path)) {
        php_error_docref(NULL, E_ERROR, "Could not resolve absolute file path");
        RETURN_FALSE;
    }

    fptr = fopen(resolved_path, "rb");
    if (fptr == NULL) {
        php_error_docref(NULL, E_ERROR, "Failed to open stream");
        RETURN_FALSE;
    }

    s_pid = play_audio(fptr, max_channels, curious, reverb);

    if (s_pid > 0) {
        MODPLAYER_G(pid) = s_pid;
    }

    RETURN_LONG(s_pid);
}
/* }}} */

/* {{{ proto int modplayer_getpid()
   Get the PID of the player sub-process, if the player is not running returns FALSE. */
PHP_FUNCTION(modplayer_getpid)
{
    if (MODPLAYER_G(pid) > 0) {
        RETURN_LONG(MODPLAYER_G(pid));
    } else {
        RETURN_FALSE;
    }
}
/* }}} */

/* {{{ proto bool modplayer_stop()
   Stop the module player.*/
PHP_FUNCTION(modplayer_stop)
{
    pid_t m_pid;

    m_pid = MODPLAYER_G(pid);

    if (m_pid > 0) {
        kill(m_pid, SIGKILL);
        MODPLAYER_G(pid) = 0;

        RETURN_TRUE;
    }

    php_error_docref(NULL, E_WARNING, "The player module was not started yet.");
    RETURN_FALSE;
}
/* }}} */
